bindings/pyroot/pythonizations/python/ROOT/_pythonization/_function_utils.py

```diff
@@ -0,0 +1,224 @@
+        """
+        Function to determine the type of a variable x.
+        If x is a string. It maps it to the corresponding column's type in RDF. (String constants are not supported)
+        Else it determines the fundemental type of x. (int, float, double)
+        Else it is an a numpy array and maps it to corresponding RVec.
+        Otherwise falgs a type error
+        Args:
+	    `   x: Variable whose type is to be determined
+
+        Returns:
+            type of x
+        """
+        try:
+            import numpy as np
+        except:
+            raise ImportError("Failed to import numpy during call to determine function signature.")
+        if isinstance(x, str):
+            #Can be string constant or can be column name
+            if x in self.col_names: # If x is a column
+                t = self.rdf.GetColumnType(x)
+                if t in map_tree_to_numbadeclares.keys(): # The column is a fundemental type from tree  
+                    return map_tree_to_numbadeclares[t]
+                elif '<' in  t: # The column type is a RVec<type>
+                    g = re.match('(.*)<(.*)>', t).groups(0)
+                    if g[1] in map_tree_to_numbadeclares.keys():
+                        return "RVec<" + map_tree_to_numbadeclares[g[1]] + ">"
+                    return "RVec<" + str(g[1]) + ">" # There are data type that leak into here. NOt sure from where. BUt need to implement something here.
+                    
+                else:
+                    return t
+            else :
+                return 'str'
+                #! Numba Declare does not support "string" type. Check _numbadeclare.Thus, Cannot pass string constants to the filter/Defines..
+        elif isinstance(x, bool):
+            return 'bool'
+        elif isinstance(x, int):
+            return 'int'
+        elif isinstance(x, float):
+            return 'float'
+        elif isinstance(x, np.ndarray): 
+            if x.dtype.type in  map_numpy_to_tree:
+                return "RVec<" + map_numpy_to_tree[x.dtype.type] + ">"
+            else:
+                raise TypeError(f"Support for {x.dtype.type} arrays is not yet supported.")
+        #! Need to work out how to map things like tuples, dicts, lists...
+        else:
+                raise TypeError(f"Type of {type(x)}:  {x} is not supported to be jitted.")
+            
+    def find_function_signature(self, func, funcargs):
+        """
+        Function to determine the function signature i.e the type of variables the function arguments and its corresponding value.
+        Args:
+	    `   func: Function whose signature is to be determined.
+                    P.S The function should not be crazy. It should have simple datatypes for the arguments and should not have default arguments. (Use funcargs instead)
+            funcargs: Dictionatry containing constant parameters the that func requires. Or any special mapping of the arguments of func.
```

Comment:
I'm not convinced we should provide this parameter for passing default arguments.  Does it work if the python callable captures variables from the scope where it is defined? If yes, the callable could be of the form
```
a = 1
def my_function(col_name1, col_name2, def_arg1 = val1, def_arg2 = val2):
    # "a" can be used here 
    # `my_function` will be always called by RDataFrame just passing `col_name1` and `col_name2` as args
    ...
```
Having default args like this might not make a lot of sense, but it could be supported.

Reply:
Yes. I think the feature was based off the tutorial in pyroo004. Where we used a pypow(x,y) function. In this case y was hard coded as 3 and 2 in the C++ call string. I just wanted to provide a pythonic way to do it. Thus a way for the user to pass constant arguments.
Yes they can use the variable from the scope. The type would still be recognised as int and the call would use the value 1. Acc to your case.

Reply:
In your example from above the kwargs `def_arg1` and `def_arg2` can only take their default value, there is no way of passing another value to those two kwargs. I think either we only allow column names and the users can write closures using outer-scope variables inside the function scope, or we have to support kwargs for the jitted function being passed to the Filter

Reply:
Although, there might be a caveat here. The use of the variable "a" inside the function. Which would, later on, be jitted might cause "a" to be considered as a constant by numba. Thus we might hinder the modification of "a" inside the function by the user.